Content.Shared/Materials/SharedMaterialStorageSystem.cs

```diff
@@ -0,0 +1,227 @@
+            return 0; //you have nothing
+        return !component.Storage.TryGetValue(material, out var amount) ? 0 : amount;
+    }
+
+    /// <summary>
+    /// Gets the total volume of all materials in the storage.
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="component"></param>
+    /// <returns>The volume of all materials in the storage</returns>
+    public int GetTotalMaterialAmount(EntityUid uid, MaterialStorageComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return 0;
+        return component.Storage.Values.Sum();
+    }
+
+    /// <summary>
+    /// Tests if a specific amount of volume will fit in the storage.
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="volume"></param>
+    /// <param name="component"></param>
+    /// <returns>If the specified volume will fit</returns>
+    public bool CanTakeVolume(EntityUid uid, int volume, MaterialStorageComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+        return component.StorageLimit == null || GetTotalMaterialAmount(uid, component) + volume <= component.StorageLimit;
+    }
+
+    /// <summary>
+    /// Checks if the specified material can be changed by the specified volume.
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="materialId"></param>
+    /// <param name="volume"></param>
+    /// <param name="component"></param>
+    /// <returns>If the amount can be changed</returns>
+    public bool CanChangeMaterialAmount(EntityUid uid, string materialId, int volume, MaterialStorageComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+        return CanTakeVolume(uid, volume, component) &&
+               (component.MaterialWhiteList == null || component.MaterialWhiteList.Contains(materialId)) &&
+               (!component.Storage.TryGetValue(materialId, out var amount) || amount + volume >= 0);
+    }
+
+    /// <summary>
+    /// Changes the amount of a specific material in the storage.
+    /// Still respects the filters in place.
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="materialId"></param>
+    /// <param name="volume"></param>
+    /// <param name="component"></param>
+    /// <returns>If it was successful</returns>
+    public bool TryChangeMaterialAmount(EntityUid uid, string materialId, int volume, MaterialStorageComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+        if (!CanChangeMaterialAmount(uid, materialId, volume, component))
+            return false;
+        if (!component.Storage.ContainsKey(materialId))
+            component.Storage.Add(materialId, 0);
+        component.Storage[materialId] += volume;
+
+        if (!_defferedMaterialChangeEvent.Contains(component))
+            _defferedMaterialChangeEvent.Add(component);
```

Comment:
Do these need to be deferred? Storing data on systems become funny with serialization.

Reply:
this was done to cut down on a few extra ui messages and updates, but yea, it probably doesn't matter and it was done like that before.